C API 2

jkankiewicz opened the pull request:
@alexjg, @ConradIrwin, and @orionz, this PR is my response to @ConradIrwin's issue #451.

The `AMvalue` union,  `AMlistItem` struct, `AMmapItem` struct, and `AMobjItem` struct are gone, replaced by the `AMitem` struct.

The `AMchangeHashes`, `AMchanges`, `AMlistItems`, `AMmapItems`, `AMobjItems`, `AMstrs`, and `AMsyncHaves` iterators are gone, replaced by the `AMitems` iterator.

The `AMitem` struct is opaque, getting and setting values is now achieved exclusively through function calls.

The `AMitemsNext()`, `AMitemsPrev()`, and `AMresultItem()` functions return a pointer to an `AMitem` struct so you ultimately get the same thing whether you're iterating over a sequence or calling `AMmapGet()` or `AMlistGet()`.

Calling `AMitemResult()` on an `AMitem` struct will produce a new `AMresult` struct referencing its storage so now the `AMresult` struct for an iterator can be subsequently freed without affecting the `AMitem` structs that were filtered out of it.

The storage for a set of `AMitem` structs can be recombined into a single `AMresult` struct by passing pointers to their corresponding `AMresult` structs to `AMresultCat()`.

I hope that these changes will enable the implementation of the automerge-go library to be simplified.

For C/C++ programmers, I've added `AMstrCmp()`, `AMstrdup()`, `AM{idxType,objType,status,valType}ToString()` and `AM{idxType,objType,status,valType}FromString()`. It's also now possible to pass arbitrary parameters through `AMstack{Item,Items,Result}()` to a callback function.

ConradIrwin commented:
I took a quick look at this this evening to see if I could get some directional feedback!

Overall: I like the reduced surface area by having one iterator type, and it makes a lot of sense to me. I can get rid of the wrapper functions I added, and am able to call all the functions I want to from go, which is great!

A few suggestions (hastily made, so feel free to ignore):
* I would love if `AMitemsNext/Prev` just returned an `AMresult` – currently that's the only function that returns an item directly so I have to wrap it specially.
* Another small quality of life improvement would be to make the `AMitemToX` call have a signature that can tell you if you guessed the type correctly. e.g. `int AMitemToActorId(const struct AMitem *item, struct AMactorId** result)`, so I could combine the type-check and the value extraction into one line. (This obviously doesn't work for AMmapGet, etc.; but for many functions the type is known ahead of time).
* A really minor thing: `AMcommit` returns a change hash if it was successful or a void otherwise. It would be better to either return an explicit error that I could handle, or return a result with AMsize() == 0. That way I could just pass the result into the same code that handles getting a list of changesHashes that I needed to write for `AMheads()` instead of having a separate function.
* (a bit of a crazy idea) should AMitem and AMresult be merged into one? It's not clear why they are separate (from an API users point of view) as I always have to first check "does the result have an error, and does it have just one item, and is the items type correct, if so, get the item". It would be really nice if errors were designated by a different type, so I could reduce the checks to "does this itemresult have the right type?". 

I'm looking forward to updating automerge-go based on this!

ConradIrwin commented:
I took a quick look at this this evening to see if I could get some directional feedback!

Overall: I like the reduced surface area by having one iterator type, and it makes a lot of sense to me. I can get rid of the wrapper functions I added, and am able to call all the functions I want to from go, which is great!

A few suggestions (hastily made, so feel free to ignore):
* I would love if `AMitemsNext/Prev` just returned an `AMresult` – currently that's the only function that returns an item directly so I have to wrap it specially.
* Another small quality of life improvement would be to make the `AMitemToX` call have a signature that can tell you if you guessed the type correctly. e.g. `int AMitemToActorId(const struct AMitem *item, struct AMactorId** result)`, so I could combine the type-check and the value extraction into one line. (This obviously doesn't work for AMmapGet, etc.; but for many functions the type is known ahead of time).
* A really minor thing: `AMcommit` returns a change hash if it was successful or a void otherwise. It would be better to either return an explicit error that I could handle, or return a result with AMsize() == 0. That way I could just pass the result into the same code that handles getting a list of changesHashes that I needed to write for `AMheads()` instead of having a separate function.
* (a bit of a crazy idea) should AMitem and AMresult be merged into one? It's not clear why they are separate (from an API users point of view) as I always have to first check "does the result have an error, and does it have just one item, and is the items type correct, if so, get the item". It would be really nice if errors were designated by a different type, so I could reduce the checks to "does this itemresult have the right type?". 

I'm looking forward to updating automerge-go based on this, and will keep feedback coming as I notice things.

jkankiewicz commented:
> I took a quick look at this this evening to see if I could get some directional feedback!

@ConradIrwin, thank you! Constructive feedback on the C API is always welcome.

> I would love if AMitemsNext/Prev just returned an AMresult – currently that's the only function that returns an item directly so I have to wrap it specially.

I believe that always creating a new `AMresult` when calling `AMitemsNext()` or `AMitemsPrev()` would be overkill for a couple of reasons:

1. The returned `AMresult` struct would never contain an error message; the `AMresult` struct from which the `AMitems` struct was extracted would already have contained an error message if something had gone wrong.
2. Not every `AMitem` struct will be selected for retention whilst iterating; it's more efficient to only pass an `AMitem` pointer to `AMitemResult()` when an `AMitem` struct should outlive the `AMitems` struct from which it was extracted.
You may choose to always call `AMitemResult(AMitemsNext(...))` instead of `AMitemsNext(...)` or `AMitemResult(AMitemsPrev(...))` instead of `AMitemsPrev(...)` but the cost of that choice shouldn't be forced upon everyone all of the time.

> Another small quality of life improvement would be to make the AMitemToX call have a signature that can tell you if you guessed the type correctly. e.g. int AMitemToActorId(const struct AMitem *item, struct AMactorId** result), so I could combine the type-check and the value extraction into one line. (This obviously doesn't work for AMmapGet, etc.; but for many functions the type is known ahead of time).

I had considered returning the value of an `AMitem` struct through an out parameter because doing so avoids the need to designate an arbitrary, null-equivalent value when it's not a pointer e.g. `AMitemToUint()` returns `SIZE_MAX` when its `AMitem const*` argument is `NULL`. I chose instead to return the value directly only because the `AMresultValue()` function that the `AMitemTo...()` functions replace works that way but you've changed my mind.
This PR is already undergoing a review so I'll open a new PR wherein the `AMitemTo...()` functions return their respective types of value through an out parameter instead.

> A really minor thing: AMcommit returns a change hash if it was successful or a void otherwise. It would be better to either return an explicit error that I could handle, or return a result with AMsize() == 0. That way I could just pass the result into the same code that handles getting a list of changesHashes that I needed to write for AMheads() instead of having a separate function.

As @alexjg pointed out in #451,
> The return type in Rust is `Option<ChangeHash>`

It works the way it does because `Option<ChangeHash>::None` is not an error and the C API signifies an empty result with a single `AMitem` struct whose type is `AM_VAL_TYPE_VOID`;  there must exist an `AMitem` pointer to pass into `AMitemValType()`. Therefore, the C API can't return an empty list of `AM_VAL_TYPE_CHANGE_HASH` because it never returns an empty list.
It could return a single `AMitem` struct of type `AM_VAL_TYPE_CHANGE_HASH` with a special value, say a zeroed change hash, but doing so for just that case would be inconsistent with the rest of the API.

Have you considered always passing the `AMresult` pointer from an `AMcommit()` call to `AMresultItems()`?
When there are no change hashes in the result that call will yield a zeroed `AMitems` struct which signifies an empty iterator i.e. passing the pointer for a zeroed `AMitems` struct to `AMitemsSize()` will always yield `0` and passing it to `AMitemsNext()` or `AMitemsPrev()` will always yield `NULL`. Could doing that work for you?

> (a bit of a crazy idea) should AMitem and AMresult be merged into one? It's not clear why they are separate (from an API users point of view) as I always have to first check "does the result have an error, and does it have just one item, and is the items type correct, if so, get the item". It would be really nice if errors were designated by a different type, so I could reduce the checks to "does this itemresult have the right type?".

I'll just paraphrase the same answer that I gave @rkuhn when he proposed the same thing in [#411](https://github.com/automerge/automerge/issues/411#issuecomment-1191634790):

The C API's functions don't take `AMresult` arguments in order to leverage static type checking by the C compiler and to allow them to directly return primitive values such as the length of a vector.
If a C API function were to take `AMresult` arguments then it would have to check if any of them contains an error message or an unexpected type of value and it would also have to wrap its result value, even a primitive one, in an `AMresult` just so that it could report that one of its arguments was an error message or a value of an unexpected type.
The bifurcation of `AMresult` and `AMitem` encourages error detection to take place only once (when the `AMitem` pointer  or `AMitems` struct is extracted) and within the context where the error can actually be handled and not just forwarded.

> I'm looking forward to updating automerge-go based on this, and will keep feedback coming as I notice things.

That's great to hear although you may wish to wait until after the `AMitemTo...()` functions return a value through an out parameter.

ConradIrwin commented:
Thanks for the detailed responses, your approach makes sense to me!

The main one I'm still not sure on is `AMcommit`, there are two "rules of thumb" I typically use when designing APIs:
* When you call a function you should be able to tell based on the input types what the output types are.
* When you call a function and it cannot do what you expect, it should fail loudly.

In the rust version this is not a problem, because the compiler tells you about the optional. In the C version, there's no way to tell (well, I could have read the comment above the function, but I didn't because by the time I was wrapping AMcommit I'd already wrapped a bunch of functions that didn't have this gotcha).

It's not a huge problem because I can detect this case and expose a more go-feeling API to go programmers; but it was something that tripped me up when first building it. (particularly because the case that commit doesn't commit is pretty edge-casey).

Although it's different from a rust API, I would expect a C API to return an error and not a void.

(The other similar function is AMchange, which returns VOID on change-not-found).

Further random observations:
* I am going to be able to end up with significantly less code overall, which is great!
* I really like that with AMitemToObjId I no longer need to retain the item.
* I like the consistency with passing AMitems for all multi-valued arguments (though it took a few minutes of figuring out to find AMresultCat to create AMitems).
* `AMfree` and `AMerrorMessage` might be better named `AMresultFree` and `AMresultError` so that they follow the pattern of every method starting with the name of the type it operates on.

ConradIrwin commented:
Oh, one other thing; it looks like this now builds `lib_automerge.a` for me, instead of `libautomerge.a`. Not sure if that was intended, but I think most libraries don't hvae the underscore in the name.

jkankiewicz commented:
> Oh, one other thing; it looks like this now builds lib_automerge.a for me, instead of libautomerge.a. Not sure if that was intended, but I think most libraries don't hvae the underscore in the name.

> Not sure if that was intended, but I think most libraries don't hvae the underscore in the name.

@ConradIrwin, it has an underscore in its name because lib_automerge.a is not the entire library, it's just the C bindings for the underlying Rust library. Some of the structs and functions are implemented in C now so you should be linking against libautomerge.a instead.
You may have noticed that I added installation instructions to the README.md file and I did this because the library's artifacts should be found in their installed locations and not within the build directory.

jkankiewicz commented:
> I like the consistency with passing AMitems for all multi-valued arguments...

@ConradIrwin, you can also use `AMresultFrom()` which is better for concatenating an arbitrary number of `AMresult` structs; it's one of the functions that are implemented in C.

> ...(though it took a few minutes of figuring out to find AMresultCat to create AMitems).

It may be easier for you to find what you need by reading the HTML documentation because it groups related functions declared in different header files, like `AMresultCat()` and `AMresultFrom()`, within the same page.

I added instructions for building the documentation to the README.md file.

jkankiewicz commented:
> I really like that with AMitemToObjId I no longer need to retain the item.

@ConradIrwin, you do still need to retain the `AMresult` struct for the `AMitem` struct whose pointer you passed to `AMitemObjId()` though.

jkankiewicz commented:
> I really like that with AMitemToObjId I no longer need to retain the item.

Just to be clear for others who may be reading this: being that you **do** have to retain the `AMresult` that stores the `AMitem` from which the `AMitemObjId()` call extracted the `AMobjId` pointer, you're still implicitly retaining the `AMitem`.

jkankiewicz commented:
> AMfree and AMerrorMessage might be better named AMresultFree and AMresultError so that they follow the pattern of every method starting with the name of the type it operates on.

@ConradIrwin, those function names were inherited from the initial implementation of the C API by @orionz.

@orionz, what do you think of his suggestion?

ConradIrwin commented:
Good to know!

Sorry for the AMobjObjId confusion - I had misread the return type and not tried it yet.

I do need to pit the built headers and shared libraries in the automerge-go repository so that go can find them - when I was trying to fix my build script I must have picked up the wrong one, will fix!

Thanks for the pointer on the HTML docs, I’ll see if I can get them building locally for now, and I’d be happy to figure out how to get them hosted on the web once this is merged so they’re easy for people to find.

ConradIrwin commented:
@jkankiewicz Looking into the changes to move some of the implementation to C I have two concerns:

* Currently I can cross-compile using `cross`, which does not require me to install C compilers on my machine (it handles setting up docker images with the rust toolchain configured appropriately automatically). Although I could try and figure out how to make CMake invoke cross in a cross-platform way, in order to restore cross-compilation support I think I'd also need to install a bunch of C compilers on my system manually (which I'm sure is possible, it just seems a bit tedious).
* As I'm using `cgo` I can't call a varargs function like `AMresultFrom` anyway :D.

If you're sure you want to go down this path I think that's ok; I can just continue to use `lib_automerge` for the go version until we have time to figure out fixing cross compilation and publishing the built libraries somewhere; but it might be a bit easier for everyone if this was not the approach we take.

I do realize that it's a bit unusual to actually want to do cross-compilation, so let me explain why it's set up the way it is:

* Go supports building modules with cgo, and it can link libraries/headers from arbitrary paths.
* Go also has really good support for building directly from a git repository (a "module" in go parlance)
* Given that, popular go modules that wrap bespoke C libraries (e.g. https://github.com/rogchap/v8go) include the pre-built libraries for supported platforms in the git repository.

This has two advantages:
* Anyone who wants to use automerge-go can just do so using the go toolchain, they don't have to have all of the automerge toolchain (rust, cmake, etc.) installed and working.
* The linked version of automerge is the version that the library was tested with; so there's no chance for version skew.

You can see how I solved this in automerge-go so far [here](https://github.com/automerge/automerge-go/tree/main/deps):

* rebuild.sh rebuilds all the shared libraries for each supported platform, and outputs it to the directory. (note in particular that I only invoke cmake to generate automerge.h; I build the libraries for all platforms with cargo/cross).
* test_all.sh runs the tests in a vanilla go-enabled docker image to show that it works
* automerge.h is shared (but that's ok, because all the platforms I currently support are 64-bit, and we can cross the 32-bit bridge if anyone really wants that – when I tried I couldn't get qemu to run go in 32-bit mode anyway :D).

As I said, it doesn't have to be a blocker; but this is a step backwards in terms of usability for me at least.

ConradIrwin commented on rust/automerge-c/src/utils/string.c:
```diff
@@ -0,0 +1,51 @@
+#include <automerge-c/utils/string.h>
+
+int AMstrcmp(AMbyteSpan const lhs, AMbyteSpan const rhs) {
+    if (lhs.count < rhs.count) {
```
I noticed this just because I was clicking around in the C files, but I think this incorrectly sorts "z" before "aa".

ConradIrwin commented:
Thanks for the changes to AMitemToX, I have just got reading and writing to maps working with the latest commit on this branch (still need to do lists,text,counters, but should be very quick). One further question:

* Should AMmapPutObject return an item with AM_VAL_TYPE_OBJ_TYPE? Currently it returns an item with AM_VAL_TYPE_VOID, but calling AMitemObjId works (in the similar case of using AMmapGet to retrive a nested map, it does return AM_VAL_TYPE_OBJ_TYPE).

jkankiewicz commented:
> [T]he changes to move some of the implementation to C...is a step backwards in terms of usability for me at least. ... I can just continue to use lib_automerge for the go version...

@ConradIrwin, I've renamed the core library built by Cargo from "_automerge" to "automerge_core" and CMake is now installing it alongside the "automerge" library so that you won't have to find it within the out-of-source build directory anymore.
I've also ported `AMstrcmp()` from C to Rust and moved it into the core library so that automerge-go can use it.

> As I'm using cgo I can't call a varargs function like AMresultFrom anyway :D.

Let me know if automerge-go requires any of the other utility functions that are written in C.

jkankiewicz commented:
> Should AMmapPutObject return an item with AM_VAL_TYPE_OBJ_TYPE?

@ConradIrwin, it does that now.

jkankiewicz commented on rust/automerge-c/src/doc/list.rs:
```diff
@@ -525,38 +537,49 @@ pub unsafe extern "C" fn AMlistPutStr(
     let obj_id = to_obj_id!(obj_id);
-    let (index, insert) = adjust!(index, insert, doc.length(obj_id));
+    let _is_max = pos == usize::MAX;
+    // let (pos, insert) = adjust!(pos, insert, doc.length(obj_id));
```
> nit: can we remove this commented stuff and the _is_max variable?

@alexjg, thanks! That should have been removed after debugging.

jkankiewicz commented on rust/automerge-c/src/item.rs:
```diff
@@ -0,0 +1,1911 @@
+
+        match value {
+            Change(change, c_change) => unsafe {
+                Ok((*c_change.get()).get_or_insert(AMchange::new(change)))
```
> How is the validity of this ensured? I think it's that the only time we create this variant is via impl From<am::Change> for Value, where we set the c_change: UnsafeCell<Option<AMChange>> to Default and whilst the Value enum is public, we don't actually expose a C API which provides access to the variants so there's no way for the user to affect the c_change pointer, is that correct?

@alexjg, the user can't affect the `c_change` field directly, no, only indirectly the first time that they retrieve an `AMchange` struct from it.

orionz commented:
> 

I have no attachment to the old names.  If these new names are more consistent they have my support.

ConradIrwin commented:
Thanks! 

No it doesn't because I don't need the result stack (I use go's GC instead) and I'm using AMresultCat instead of AMresultFrom. I also immediately convert returned AMbyteSpan into go strings or byte-slices (both of which can contain null bytes).
